diff --git a/bfd/ChangeLog b/bfd/ChangeLog
index 9fac12538a..7608f9b0ba 100644
--- a/bfd/ChangeLog
+++ b/bfd/ChangeLog
@@ -1,3 +1,7 @@
+2020-09-19  Nick Clifton  <nickc@redhat.com>
+
+	* development.sh (development): Set to true.
+
 2020-09-19  Nick Clifton  <nickc@redhat.com>
 
 	This is the 2.35.1 point release.
diff --git a/bfd/development.sh b/bfd/development.sh
index 32be4b9460..6bbed41d6d 100644
--- a/bfd/development.sh
+++ b/bfd/development.sh
@@ -16,7 +16,7 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 # Controls whether to enable development-mode features by default.
-development=false
+development=true
 
 # Indicate whether this is a release branch.
 experimental=false
