diff --git a/bfd/ChangeLog b/bfd/ChangeLog
index 36cb81dd7f3..8d6485f5450 100644
--- a/bfd/ChangeLog
+++ b/bfd/ChangeLog
@@ -1,3 +1,8 @@
+2022-02-09  Nick Clifton  <nickc@redhat.com>
+
+	* development.sh (development): Set to true.
+	* version.m4 (BFD_VERSION): Set to 2.38.0.
+
 2022-02-09  Nick Clifton  <nickc@redhat.com>
 
 	2.38 release.
diff --git a/bfd/version.m4 b/bfd/version.m4
index f611cacaa29..1b8f4a3cbaa 100644
--- a/bfd/version.m4
+++ b/bfd/version.m4
@@ -1 +1 @@
-m4_define([BFD_VERSION], [2.38])
+m4_define([BFD_VERSION], [2.38.0])
