From c27a83d95b89d1861184379dadb929b6d83dd6fd Mon Sep 17 00:00:00 2001
From: George T Kramer <george.t.kramer@intel.com>
Date: Fri, 26 May 2017 13:57:38 -0700
Subject: [PATCH] Switch build flow to production

---
 bfd/development.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bfd/development.sh b/bfd/development.sh
index cd31410..b001a88 100644
--- a/bfd/development.sh
+++ b/bfd/development.sh
@@ -16,4 +16,4 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 # Controls whether to enable development-mode features by default.
-development=true
+development=false
-- 
2.13.0

